internal/layer2/announcer.go

```diff
@@ -192,51 +204,89 @@ func (a *Announce) gratuitous(ip net.IP) {
 
 	if ip.To4() != nil {
 		for _, client := range a.arps {
+			if !isMatchIntf(client.intf, ipA.interfaces) {
+				level.Debug(a.logger).Log("op", "announcer", "info", "skip interfaces", client.intf)
+				continue
+			}
 			if err := client.Gratuitous(ip); err != nil {
 				level.Error(a.logger).Log("op", "gratuitousAnnounce", "error", err, "ip", ip, "msg", "failed to make gratuitous ARP announcement")
 			}
 		}
 	} else {
 		for _, client := range a.ndps {
+			if !isMatchIntf(client.intf, ipA.interfaces) {
+				level.Debug(a.logger).Log("op", "announcer", "info", "skip interfaces", client.intf)
+				continue
+			}
 			if err := client.Gratuitous(ip); err != nil {
 				level.Error(a.logger).Log("op", "gratuitousAnnounce", "error", err, "ip", ip, "msg", "failed to make gratuitous NDP announcement")
 			}
 		}
 	}
 }
 
-func (a *Announce) shouldAnnounce(ip net.IP) dropReason {
+func isMatchIntf(intf string, ifs sets.String) bool {
+	// Don't specify LB IP related interfaces
+	if ifs.Len() == 0 {
+		return true
+	}
+	return ifs.Has(intf)
+}
+
+func (a *Announce) shouldAnnounce(ip net.IP, intf string) dropReason {
 	a.RLock()
 	defer a.RUnlock()
 	for _, ips := range a.ips {
 		for _, i := range ips {
-			if i.Equal(ip) {
+			if i.ip.Equal(ip) {
+				if !isMatchIntf(intf, i.interfaces) {
+					return dropReasonNotMatchInterface
+				}
 				return dropReasonNone
 			}
 		}
 	}
 	return dropReasonAnnounceIP
 }
 
+// updateIPs update announce.ips depend on service info.
```

Comment:
"and returns true if the announcement has changed"

Reply:
Done